### PR TITLE
[SPARK-8850] [SQL] Enable Unsafe mode by default

### PR DESCRIPTION
**Pull request by JoshRosen**
This pull request enables Unsafe mode by default in Spark SQL. In order to do this, we had to fix a number of small issues:

**List of fixed blockers**:
- [x] Make some default buffer sizes configurable so that HiveCompatibilitySuite can run properly (#7741).
- [x] Memory leak on grouped aggregation of empty input (fixed by #7560 to fix this)
- [x] Update planner to also check whether codegen is enabled before planning unsafe operators.
- [x] Investigate failing HiveThriftBinaryServerSuite test.  This turns out to be caused by a ClassCastException that occurs when Exchange tries to apply an interpreted RowOrdering to an UnsafeRow when range partitioning an RDD.  This could be fixed by #7408, but a shorter-term fix is to just skip the Unsafe exchange path when RangePartitioner is used.
- [x] Memory leak exceptions masking exceptions that actually caused tasks to fail (will be fixed by #7603).
- [x]  ~~https://issues.apache.org/jira/browse/SPARK-9162, to implement code generation for ScalaUDF.  This is necessary for `UDFSuite` to pass.  For now, I've just ignored this test in order to try to find other problems while we wait for a fix.~~ This is no longer necessary as of #7682.
- [x] Memory leaks from Limit after UnsafeExternalSort cause the memory leak detector to fail tests. This is a huge problem in the HiveCompatibilitySuite (fixed by f4ac642a4e5b2a7931c5e04e086bb10e263b1db6).
- [x] Tests in `AggregationQuerySuite` are failing due to NaN-handling issues in UnsafeRow, which were fixed in #7736.
- [x] `org.apache.spark.sql.ColumnExpressionSuite.rand` needs to be updated so that the planner check also matches `TungstenProject`.
- [x] After having lowered the buffer sizes to 4MB so that most of HiveCompatibilitySuite runs:
  - [x] Wrong answer in `join_1to1` (fixed by #7680)
  - [x] Wrong answer in `join_nulls` (fixed by #7680)
  - [x] Managed memory OOM / leak in `lateral_view`
  - [x] Seems to hang indefinitely in `partcols1`.  This might be a deadlock in script transformation or a bug in error-handling code? The hang was fixed by #7710.
  - [x] Error while freeing memory in `partcols1`: will be fixed by #7734.
- [x] After fixing the `partcols1` hang, it appears that a number of later tests have issues as well.
- [x] Fix thread-safety bug in codegen fallback expression evaluation (#7759).

### CODE REVIEW

**Comment by JoshRosen** [comment]
Found an interesting problem where the "exception on memory leak" configuration can lead to task failure exceptions being masked.  Going to push a fix to this branch, which I may end up spinning off into a separate PR depending on how things go.

**Comment by JoshRosen** [comment]
Looks like there’s a problem with using ScalaUDFs in the unsafe GeneratedAggregate path.  It looks like there’s a case where we end up using a BoundReference expression to extract fields before passing them to the UDF.  This ends up calling the generic Row.apply(), which returns data of the wrong type when called on an UnsafeRow (specifically, the failing test tries to get a Long column but ends up getting a byte array back).

One way to fix this would be to implement code generation for ScalaUDF so that the field access expressions are code generated.

**Comment by JoshRosen** [comment]
In this latest round of tests, it looks like HiveThriftBinaryServerSuite failed a test:

```
[info] - test multiple session *** FAILED *** (2 seconds, 488 milliseconds)
[info]   java.sql.SQLException: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: sort, tree:
[info] UnsafeExternalSort [KEY#270 DESC], true, 0
[info]  ConvertToSafe
[info]   Exchange rangepartitioning(KEY#270 DESC)
[info]    UnsafeExternalSort [key#270 DESC], true, 0
[info]     Exchange rangepartitioning(key#270 DESC)
[info]      HiveTableScan [key#270], (MetastoreRelation default, test_map, None)
[info]   at org.apache.hive.jdbc.HiveStatement.execute(HiveStatement.java:275)
[info]   at org.apache.hive.jdbc.HiveStatement.executeQuery(HiveStatement.java:355)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftBinaryServerSuite$$anonfun$8$$anonfun$apply$mcV$sp$13.apply(HiveThriftServer2Suites.scala:327)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftBinaryServerSuite$$anonfun$8$$anonfun$apply$mcV$sp$13.apply(HiveThriftServer2Suites.scala:324)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftJdbcTest$$anonfun$withMultipleConnectionJdbcStatement$1.apply(HiveThriftServer2Suites.scala:438)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftJdbcTest$$anonfun$withMultipleConnectionJdbcStatement$1.apply(HiveThriftServer2Suites.scala:438)
[info]   at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
[info]   at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftJdbcTest.withMultipleConnectionJdbcStatement(HiveThriftServer2Suites.scala:438)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftBinaryServerSuite$$anonfun$8.apply$mcV$sp(HiveThriftServer2Suites.scala:209)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftBinaryServerSuite$$anonfun$8.apply(HiveThriftServer2Suites.scala:204)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftBinaryServerSuite$$anonfun$8.apply(HiveThriftServer2Suites.scala:204)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:42)
[info]   at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuite.runTest(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
[info]   at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1424)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
[info]   at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftServer2Test.org$scalatest$BeforeAndAfterAll$$super$run(HiveThriftServer2Suites.scala:450)
[info]   at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:257)
[info]   at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:256)
[info]   at org.apache.spark.sql.hive.thriftserver.HiveThriftServer2Test.run(HiveThriftServer2Suites.scala:450)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:462)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:671)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:294)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:284)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:262)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
[info]   at java.lang.Thread.run(Thread.java:745)
```

It looks like there may also be a planner bug that occurs when `unsafe` is enabled while `codegen` is disabled:

```
org.apache.spark.sql.hive.execution.HiveWindowFunctionQueryWithoutCodeGenSuite. testMultipleRangeWindows
 Failed to execute query using catalyst: Error: assertion failed: UnsafeExternalSort requires code generation to be enabled java.lang.AssertionError: assertion failed: UnsafeExternalSort requires code generation to be enabled  at scala.Predef$.assert(Predef.scala:179)  at org.apache.spark.sql.execution.UnsafeExternalSort$$anonfun$doExecute$7.apply(basicOperators.scala:281)  at org.apache.spark.sql.execution.UnsafeExternalSort$$anonfun$doExecute$7.apply(basicOperators.scala:280)  at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:48)  at 
```

This is probably easy to fix by adding configuration validation which does not allow `unsafe` to be enabled unless codegen is also enabled, then updating the relevant non-code-gen suites to also disable unsafe.  We can also modify the planner to guard against this by also checking whether codegen is enabled when figuring out whether an unsafe sort operator can be used.

Finally, it looks like a number of the HiveCompatibilitySuite tests are failing due to the minimum buffer size issue, which I'll try to fix today.

**Comment by JoshRosen** [comment]
I lowered some buffers to 4MB so that the majority of HiveCompatibilitySuite runs and have found a few other test failures; see updated ticket description for more details.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala [hunk 1]
```diff
@@ -47,7 +47,12 @@ case class Exchange(newPartitioning: Partitioning, child: SparkPlan) extends Una
 
   override def canProcessSafeRows: Boolean = true
 
-  override def canProcessUnsafeRows: Boolean = true
+  override def canProcessUnsafeRows: Boolean = {
+    // Do not use the Unsafe path if we are using a RangePartitioning, since this may lead to
+    // an interpreted RowOrdering being applied to an UnsafeRow, which will lead to
+    // ClassCastExceptions at runtime. This check can be removed after SPARK-9054 is fixed.
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala]: This may be unnecessary now that we support generic getters in UnsafeRows.

**Comment by JoshRosen** [comment]
Lots of changes are temporarily merged in here while I test on top of Reynold's struct type patch.

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by JoshRosen** [comment]
Looks like two `HiveUDFSuite` tests are now failing.  I'll investigate tomorrow morning.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Nooo! This failed a PySpark test because I need to configure a lower default page size in those tests.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
(Re-testing because a flaky MLlib test may have caused a spurious failure and I want to have the Python tests run...)

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

(Just preemptively in case the first build hits a flaky streaming test)

**Comment by JoshRosen** [comment]
Jenkins retest this please

**Comment by rxin** [comment]
I've merged this!

**Pull request closed without merging**